pdns/nod.cc

```diff
@@ -119,26 +133,39 @@ void PersistentSBF::setCacheDir(const std::string& cachedir)
 bool PersistentSBF::snapshotCurrent(std::thread::id tid)
 {
   if (d_cachedir.length()) {
-    path p(d_cachedir);
-    path f(d_cachedir);
+    filesystem::path p(d_cachedir);
+    filesystem::path f(d_cachedir);
     std::stringstream ss;
     ss << d_prefix << "_" << tid;
     f /= ss.str() + "_" + std::to_string(getpid()) + "." + bf_suffix;
-    if (exists(p) && is_directory(p)) {
+    filesystem::path ftmp(filesystem::unique_path(f.string() + ".%%%%%%%%"));
+    if (filesystem::exists(p) && filesystem::is_directory(p)) {
       try {
         std::ofstream ofile;
         std::stringstream iss;
-        ofile.open(f.string(), std::ios::out | std::ios::binary);
         {
           // only lock while dumping to a stringstream
           std::lock_guard<std::mutex> lock(d_sbf_mutex);
           d_sbf.dump(iss);
         }
         // Now write it out to the file
+        ofile.open(ftmp.string(), std::ios::out | std::ios::binary);
```

Comment:
Do we care about the TOCTOU race in that case?

Reply:
What is the race?

Reply:
We generate a name, do some stuff and then open a file for writing with that name. It looks like it could have been created as a symlink to another file in the meantime, although I agree it's likely extremely difficult, depending on how truly random the result is.

Reply:
Right! The dirty secret is that there is no TOC here. `unique_path` just operates on a string.

Reply:
(So depending on how you look at it, there is no window, or the window is infinite already.)

Reply:
Doesn't `boost::filesystem::unique_path` check whether the file exists? The documentation is a bit terse.

Reply:
It doesn't - here is the code, it's short enough to read in a minute https://github.com/boostorg/filesystem/blob/6b5e38134a336b6ea777cd8a69b6ae929819db7e/src/unique_path.cpp#L234-L270

Reply:
Ugh in my mind I thought it would be a safe function, But it cannot be if it does not create the file itself.

Reply:
more complaints here - https://stackoverflow.com/questions/43946120/how-does-boostfilesystemunique-path-resolve-the-need-in-mkstemp-analogue-i

suggests `O_EXCL | O_CREAT` on the `open`, but perhaps we want to do real `mkstemp` instead?

Reply:
It looks like even C++17 doesn't offer anything better than `mkstemp`..

Reply:
There is no interface to ofstream to request `O_EXCL | O_CREAT` afaiks, so i'll go for `mkstemp`